plugins/out_kinesis_streams/kinesis_api.c

```diff
@@ -97,6 +101,62 @@ static int init_put_payload(struct flb_kinesis *ctx, struct flush *buf,
     return -1;
 }
 
+/*
+ * Count the substring number in string.
+ */
+static int count_substring(char *str1, char *str2){
+    int x = 0;
+    char *p;
+
+    do {
+        p = strstr(str1, str2);
+        if (p != NULL) {
+            str1 = p + 1;
+            x = x + 1;
+        }
+    }
+    while(p != NULL);
+    return x;
+}
+
+/*
+ * Adjust the time_key output format according to time_key_format.
+ */
+static void create_time_format_string(char *out_buff, size_t out_size,
+                                      const char *time_key_format, const struct tm *timeptr,
+                                      size_t len, struct flb_time *tms)
+{
+    char millisecond_string[millisecond_length];
+    char nanosecond_string[nanosecond_length];
+    /* calculate the length of str, replace %F to 3 digits and %L to 9 digits */
+    out_size = strlen(time_key_format)+count_substring(time_key_format, milli_string_old)+7*count_substring(time_key_format, nano_string_old);
```

Comment:
I still do not understand the math here, why 7 * the nano string but just 1 * the milli string? We can discuss if needed.

Reply:
Is 7 a special number such that it could be a constant?